### PR TITLE
Support PropertyLoadSaver Interface

### LINKED ISSUE #53: Changes by PropertyLoadSaver.Load aren't cached.
Currently, goon.Get calls methods as below
.. => goon.Get => datastore.Get => goon.putMemcache => ..
or (if it's cached)
.. => goon.Get => memcache.Get => goon.putMemcache => ..

I think this flow has a problem.
PropertyLoadSaver.Load is called in only datastore.Get.
And goon.putMemcache ignores fields which tagged as `datastore:"-"`.
Therefore the fields changed by PropertyLoadSaver.Load aren't cached.

The return of goon.Get should be same whether it's cached or not.

### PR DESCRIPTION
**Pull request by yuichi1004**
# Introduction

This PR supports PropertyLoadSaver interface.

https://cloud.google.com/appengine/docs/standard/go/datastore/reference#hdr-The_PropertyLoadSaver_Interface

This PR should resolve #53.

# Strategy

* Define serializationStatePropertyList as new cache type
  - PropertyLists are cached with serializationStatePropertyList, while others are cached with serializationStateNormal
  - This strategy simplifies the implementation than maintaining compatibility of normal cache to PropertyLoadSaver interface
  - I think most people use one of normal struct or PropertyLoadSaver interface so this works on most of the use cases
* Support fetching different cache type if possible, fallback to datastore if not
  - This strategy supports most of the case with cache efficiency
  - This strategy avoids error even if fetching different cache type is not supported

Please take a look at the code. Any comments, improvements, proposal of the different approach are welcome.

### CODE REVIEW

**Comment by xStrom** [comment]
I'll find time for review later this week.

**Comment by hamakn** [comment]
How is the progress?
I want this feature.

**Comment by xStrom** [comment]
I apologize, this got deprioritized and somewhat forgotten.

Good to know that there's active interest! I will make an effort to get this functionality tested & merged.

@yuichi1004 are you still interested in helping make this happen? There are merge conflicts right now, I think it's these two PRs that got merged after you submitted your changes: #71 & #73.

If @yuichi1004 wants to help and fixes the merge conflicts, I will review/test it. If not, it may take a bit longer but I will do the necessary work myself.

**Comment by yuichi1004** [comment]
@xStrom @hamakn  let me take care of this. I will check and fix on this weekend. Sorry for my late response.

**Comment by yuichi1004** [comment]
@xStrom @hamakn 
Conflict resolved and test works for me.

I have one failure test but I do believe this is irrelevant to this change and the bug of App Engine dev server.

```
=== RUN   TestChangeMemcacheKey
INFO     2018-04-14 07:00:02,920 devappserver2.py:105] Skipping SDK update check.
WARNING  2018-04-14 07:00:02,920 devappserver2.py:121] DEFAULT_VERSION_HOSTNAME will not be set correctly with --port=0
WARNING  2018-04-14 07:00:02,974 simple_search_stub.py:1196] Could not read search indexes from /var/folders/5j/2bmf8d2d3xgb9lmttlptxvm8p49fjs/T/appengine.testapp.01099286/search_indexes
INFO     2018-04-14 07:00:02,976 api_server.py:308] Starting API server at: http://localhost:58270
INFO     2018-04-14 07:00:03,045 dispatcher.py:255] Starting module "default" running at: http://localhost:58271
INFO     2018-04-14 07:00:03,046 admin_server.py:146] Starting admin server at: http://localhost:58273
2018/04/14 16:00:03 Metadata fetch failed: Get http://metadata/computeMetadata/v1/instance/attributes/gae_backend_version: dial tcp: lookup metadata: no such host
exit status 1
FAIL    github.com/yuichi1004/goon      1.453s
```

**Comment by hamakn** [comment]
Using `goapp test`, all tests are passed on this branch in my environment.
````
$ git branch
  master
* topic/support_property_load_saver
$ goapp test ./...
ok      github.com/mjibson/goon 209.410s
````

**Comment by yuichi1004** [comment]
@delphinus @daisuzu 
Appreciate the feedback. Let me check that in detail over the weekend. I will keep you guys updated.

**Comment by yuichi1004** [comment]
@daisuzu Well, that's a silly mistake... 7b65b06e8f6aaf560d5e61764c23e7e19ef58538
After the fix above, it looks working.

```
package goon

import (
	"testing"

	"google.golang.org/appengine/aetest"
	"google.golang.org/appengine/datastore"
	"google.golang.org/appengine/memcache"
)

type Value struct {
	ID     int64  `datastore:"-" goon:"id"`
	String string `datastore:"string"`
}

func (v *Value) Save() ([]datastore.Property, error) { return datastore.SaveStruct(v) }
func (v *Value) Load(p []datastore.Property) error   { return datastore.LoadStruct(v, p) }

func TestPLS(t *testing.T) {
	c, done, err := aetest.NewContext()
	if err != nil {
		t.Fatal(err)
	}
	defer done()
	g := FromContext(c)

	// Put
	v := Value{String: "test"}
	k, err := g.Put(&v)
	if err != nil {
		t.Fatal(err)
	}
	g.FlushLocalCache()

	// Get from datastore
	v1 := Value{ID: v.ID}
	if err := g.Get(&v1); err != nil {
		t.Fatal(err)
	}
	t.Log("from datastore:", v1)

	// Get from memcache
	s, err := memcache.Get(c, MemcacheKey(k))
	if err != nil {
		t.Fatal(err)
	}
	var v2 Value
	if err := deserializeStruct(&v2, s.Value); err != nil {
		t.Fatalf("could not deserialize `%v`: %s", s.Value, err) // err == errCacheFetchFailed
	}
	t.Log("from memcache (raw):", v2)

	v3 := Value{ID: v.ID}
	if err := g.Get(&v3); err != nil {
		t.Fatal(err)
	}
	t.Log("from memcache (thorugh goon):", v3)
}
```

```
--- PASS: TestPLS (3.98s)
        my_test.go:40: from datastore: {5629499534213120 test}
        my_test.go:51: from memcache (raw): {0 test}
        my_test.go:57: from memcache (thorugh goon): {5629499534213120 test}
```

**Comment by daisuzu** [comment]
@yuichi1004 

Ah, I see...
After 005ae6a, I confirmed it works well.
Thank you!

**Comment by yuichi1004** [comment]
@delphinus I am still looking into that topic.
I think your argument is correct. I am trying to avoid using local cache if dst is PLS since current local cache does not have serialization process which makes difficult to introduce `Save()` and `Load()`. 

We can simplify all of those complexity if we fully migrated to PropertyList based implementation. If underlying core logic were written in PropertyList, then PLS struct and non-PLS struct can work on the same caching logic.

But for now, I will go with the easy approach since this is a bigger discussion than original topic.

**Comment by delphinus** [comment]
Agree.

If it contains PLS process for local cache, it should encode entities into cache instead of the raw structs as current logic does.  It makes some performance impact, and so I also think your plan is good for this time.

**Comment by yuichi1004** [comment]
I am still working on avoiding local cache for PLS struct. I got weird error related to re-using deserialization decoder.

I appreciate if anybody can  provide the fix, otherwise I will work on next weekend.

File: entity.go [hunk 1]
```diff
@@ -391,22 +416,39 @@ func serializeStruct(src interface{}) ([]byte, error) {
...
 
-	bufSize := se.buf.Len()
-	// final size = header + all metadatas + separators for metadata + data
-	finalBufSize := 1 + smd.totalLength + len(smd.metaDatas) + bufSize
-	finalBuf := make([]byte, finalBufSize)
-	finalBuf[0] = byte(serializationStateNormal)          // Set the header
-	serializeStructMetaData(finalBuf[1:], smd)            // Serialize the metadata
-	copy(finalBuf[finalBufSize-bufSize:], se.buf.Bytes()) // Copy the actual data
+	if ls, ok := src.(datastore.PropertyLoadSaver); ok {
+		se.buf.Write([]byte{serializationStatePropertyList})
```
Comment (xStrom) [inline entity.go]: We should use the `WriteByte` method to skip the slice allocation. Also we should check for the error returned by this method.

File: goon_test.go [hunk 2]
```diff
@@ -856,6 +869,49 @@ type MigrationB struct {
...
+}
+func (m *MigrationPlsB) Save() ([]datastore.Property, error) {
+	return datastore.SaveStruct(m)
+}
+func (m *MigrationPlsB) Load(props []datastore.Property) error {
+	return datastore.LoadStruct(m, props)
+}
+
+// Make sure that MigrationC implements datastore.PropertyLoadSaver
+var _, _ datastore.PropertyLoadSaver = &MigrationPlsA{}, &MigrationPlsB{}
```
Comment (xStrom) [inline goon_test.go]: The `MigrationC` reference in the comment seems outdated. Should change the wording to something like: _Make sure that these implement datastore.PropertyLoadSaver_.

File: goon_test.go [hunk 3]
```diff
@@ -892,24 +948,64 @@ func TestMigration(t *testing.T) {
 	}
 
 	// Run migration tests with both IgnoreFieldMismatch on & off
-	for i := 0; i < 2; i++ {
-		IgnoreFieldMismatch = (i == 0)
+	testcase := []struct {
+		name                string
+		ignoreFieldMismatch bool
+		src, dst            MigrationEntity
+	}{
```
Comment (xStrom) [inline goon_test.go]: We should still test with `IgnoreFieldMismatch` on & off with all the variants, not just _NormalCache -> NormalCache_. Writing all of these out can get too verbose (8 cases), so we could still do a loop `for i := 0; i < 2; i++ {` like before and inside the loop add the 4 different src/dst combo types, with `IgnoreFieldMismatch` set based on the loop variable.

**Comment by xStrom** [comment]
I've now gone over the changes once. I pointed out some basic issues. I also have some more fundamental thoughts on the implementation strategy. I'll need a bit more time to think this over and then I'll comment on it.

**Comment by yuichi1004** [comment]
@xStrom thanks for your review and comment. For the discussion about reference-based cache over copy-based cache, I actually like reference-based cache.

My assumption was some users use `Load()` func as a hook the event. It does make sense to me for `Save()` case, for example updating timestamps on `UpdatedAt` fields. But for `Load()` case, I could not come up with the concrete situation. So may be we can skip calling `Load()` for local cache.

@delphinus correct me if you need `Load()` to be called for some specific use cases.

**Comment by delphinus** [comment]
@xStrom I understand.  This behavior is correct only when `Load()` and `Save()` is pseudo-“inverse functions”, I think.

```go
type Entity {
  ID        int64 `datastore:"-" goon:"id"`
  Foo       string
  Unix      int64
  CreatedAt time.Time `datastore:"-"`
}

func (e *Entity) Load([]datastore.Property), error) {
  _ = datastore.LoadStruct(e)
  e.CreatedAt = time.Unix(e.Unix, 0)
  return nil
}

func (e *Entity) Save() ([]datastore.Property, error) {
  e.Unix = e.CreatedAt.Unix()
  return datastore.SaveStruct(e)
}
```

The code above runs good with `goon` and the raw `datastore` because `Load()` is inverse from `Save()` and vice versa, and they have no side effects.

```go
func (e *Entity) Load([]datastore.Property), error) {
  _ = datastore.LoadStruct(e)
  e.CreatedAt = time.Unix(e.Unix, 0)
  e.Foo = e.Foo + " Loaded!!" // This breaks symmetry
  return nil
}

func (e *Entity) Save() ([]datastore.Property, error) {
  e.Unix = e.CreatedAt.Unix()
  e.Foo = e.Foo + " Saved!" // This breaks symmetry
  return datastore.SaveStruct(e)
}
```

But how about above?  They are both inverse, but have side effects.  With `datastore` package, `Get` after `Put` shows the entity has the `Foo` property is ` Loaded! Saved!`.  But with `goon`, it shows only ` Saved!` because `Load()` is not called for local cache.

<details><summary>test log (failed)</summary>

```
=== RUN   TestGetAfterPut
INFO     2018-04-25 01:05:00,749 devappserver2.py:105] Skipping SDK update check.
WARNING  2018-04-25 01:05:00,749 devappserver2.py:121] DEFAULT_VERSION_HOSTNAME will not be set correctly with --port=0
WARNING  2018-04-25 01:05:00,793 simple_search_stub.py:1196] Could not read search indexes from /var/folders/fk/xlnw71yx7yd26pqby31xgdrdtbpfqh/T/appengine
testapp.jinnouchi.yasushi/search_indexes
INFO     2018-04-25 01:05:00,796 api_server.py:265] Starting API server at: http://localhost:58832
INFO     2018-04-25 01:05:00,800 dispatcher.py:255] Starting module "default" running at: http://localhost:58833
INFO     2018-04-25 01:05:00,802 admin_server.py:152] Starting admin server at: http://localhost:58834
2018/04/25 10:05:01 DEBUG: saved e: goon.Entity{ID:5629499534213120, Foo:" Saved!", Unix:1545568496, CreatedAt:time.Time{wall:0x0, ext:63681165296, loc:(*
ime.Location)(nil)}}
2018/04/25 10:05:01 DEBUG: get entity: goon.Entity{ID:5629499534213120, Foo:" Saved!", Unix:1545568496, CreatedAt:time.Time{wall:0x0, ext:63681165296, loc
(*time.Location)(nil)}}
--- FAIL: TestGetAfterPut (3.63s)
        hoge_test.go:61:
                        Error Trace:    hoge_test.go:61
                        Error:          Not equal:
                                        expected: " Saved! Loaded!"
                                        actual  : " Saved!"
                        Test:           TestGetAfterPut
                        Messages:       Load() is called after Save()
FAIL
exit status 1
FAIL    github.com/yuichi1004/goon      3.645s
```
</details>

[test code for this][test]

[test]: https://gist.github.com/delphinus/a21b9802d9c446217c93abb997899ef1

Yes, this is not much practical.  But someone may need simply hooks there (as @yuichi1004 says) and he/she might be at a loss this behavior that differs from `datastore`'s one.

This cannot be solved easily and that “side effectable” `Load()/Save()` may be undesired from the beginning.  So I think it is good if you add comments that warns this somewhere.

**Comment by xStrom** [comment]
Thanks for the test case, it illustrates an interesting case. In my previous comment I talked about how `Load` is always executed and the results cached. However here, indeed, you have provided a case where `Load` won't be executed even once.

What happens is that `Put` populates the local cache. What we could do is check per entity if it's PLS or not. If PLS then never populate local cache with `Put`. That way the first `Get` after `Put` would go to the datastore, execute `Load` and cache the result. Additional `Get` calls would use the local cache, but that cached entity would have the at-least-once `Load` guarantee.

---

On a somewhat related topic, of course there are still cases where if you `Put` a struct, and then `Get` the same entity but into a different struct or PLS, with the same goon context, it won't work as expected. The local cache does not support any migration at all currently. There are ways to support migration, for example with a PropertyList-based cache. However this has a high memory cost and a performance cost, for a feature that almost nobody actually needs. Thus I think we would need people to come forward with complaints until we start thinking about supporting local cache migrations. This should be better documented and I'll try to get that done.

**Comment by yuichi1004** [comment]
@xStrom @delphinus Thanks for your discussion. Let me clarify the action items

- Load() never called in case @delphinus pointed out
  - I am working on the fix to avoid local cache for `PUT` if the struct is PLS
- Several comments @xStrom commented
  - I am working on the fix
- I am still waiting for @xStrom 's thought about the fundamental design

**Comment by xStrom** [comment]
Speaking broadly, I think having a unified strategy for structs/PLS is superior. It would be fewer algorithms to reason about and probably less chance for bugs to be introduced in the future when changes are made. It would also allow us to fully support migrations between types. As opposed to the current situation where we can't load a PropertyList-based cache into a struct.

Then there's still the question of which type of unified strategy to use. Depend more on the datastore package and do everything with `PropertyList` or use the goon's serialization engine for everything? There are benefits on both sides. The `PropertyList` based solution would be a lot less code, less room for us to make mistakes. The goon serialization engine however is future-compatible [1], uses less storage and is slightly faster. [2] In addition the custom serialization would allow easier support for potential complex features like supporting maps without PLS, although admittedly there is no current work going on to do this.

Intuitively I think the correct solution is a hybrid of these two strategies, though leaning more towards goon's custom serialization. However this is a non-trivial undertaking. There's a saying that perfect is the enemy of good. I think that applies here as well.

We should continue with the current split strategy that we have in this PR. We only need some smaller changes, including a few that I'll comment on soon. Then we can get this merged as an initial working version. Later on I'll open up another PR for work towards a unified solution, which can have its own schedule and won't hold back this PR.

---

[1] One scenario where just encoding `PropertyList` with `gob` would fail is if Google adds some new pointer/interface to `Property` that can be nil. `gob` fails at encoding nil values and would break goon until we release a patch.

[2] Really old goon versions actually just used `gob` to serialize structs. Some of the benefits of moving to a custom serialization engine were reduced space usage and slight performance gains. The space usage in particular is interesting, because entities designed to neatly fit the 1MB datastore limit should also fit into the 1MB memcache limit. Just encoding `Property` saves field names like "NoIndex" and "Multiple", which take extra space and can be omitted in a custom solution, with just the values saved.

File: entity.go [hunk 4]
```diff
@@ -81,6 +82,10 @@ type serializationDecoder struct {
 	dec *gob.Decoder
 }
 
+type nilType struct {
+	GoonNilValue string
+}
+
```
Comment (xStrom) [inline entity.go]: Let's cut down on allocations and create a reusable value of this: `var nilValue = nilType{}`.

File: entity.go [hunk 5]
```diff
@@ -132,6 +139,24 @@ func init() {
...
+	gob.Register(seBoot.v07)
+	gob.Register(seBoot.v08)
+	gob.Register(seBoot.v09)
+	gob.Register(seBoot.v10)
+	gob.Register(seBoot.v11)
+	gob.Register(seBoot.v12)
+	gob.Register(seBoot.v13)
+	gob.Register(seBoot.v14)
+	gob.Register(seBoot.v15)
+	gob.Register(nilType{})
```
Comment (xStrom) [inline entity.go]: Use the new `nilValue` instead of allocating a new one with `nilType{}`.

File: entity.go [hunk 6]
```diff
@@ -391,22 +416,39 @@ func serializeStruct(src interface{}) ([]byte, error) {
...
+	if ls, ok := src.(datastore.PropertyLoadSaver); ok {
+		se.buf.Write([]byte{serializationStatePropertyList})
+		props, err := ls.Save()
+		if err != nil {
+			return nil, err
+		}
+		for i := 0; i < len(props); i++ {
+			v := reflect.ValueOf(props[i].Value)
+			if v.Kind() == reflect.Ptr && v.IsNil() {
+				props[i].Value = nilType{}
```
Comment (xStrom) [inline entity.go]: Use the new `nilValue` instead of allocating a new one with `nilType{}`.

File: entity.go [hunk 7]
```diff
@@ -605,22 +651,52 @@ func deserializeStruct(dst interface{}, b []byte) error {
...
+			}
+		}
+	case serializationStatePropertyList:
+		var props []datastore.Property
+		if err := sd.dec.Decode(&props); err != nil {
+			return errCacheFetchFailed
+		}
+		for i := 0; i < len(props); i++ {
+			nilVal := nilType{}
+			if props[i].Value == nilVal {
```
Comment (xStrom) [inline entity.go]: Compare to the new `nilValue` instead of the local `nilVal`.

File: entity.go (hunk 4 above)
Comment (yuichi1004) [reply entity.go]: Fixed 0ffa7caaada1ace69f367ee36d201bcebed76bbd

File: entity.go (hunk 5 above)
Comment (yuichi1004) [reply entity.go]: Fixed 0ffa7caaada1ace69f367ee36d201bcebed76bbd

File: entity.go (hunk 6 above)
Comment (yuichi1004) [reply entity.go]: Fixed 0ffa7caaada1ace69f367ee36d201bcebed76bbd

File: entity.go (hunk 7 above)
Comment (yuichi1004) [reply entity.go]: Fixed 0ffa7caaada1ace69f367ee36d201bcebed76bbd

File: entity.go (hunk 1 above)
Comment (yuichi1004) [reply entity.go]: Fixed b5dc7dcd1dcd97975ca1b10366488c2d2a153059

File: goon_test.go (hunk 2 above)
Comment (yuichi1004) [reply goon_test.go]: Fixed dd18dfdcf507a3c9df03d209d940e71b82833e10

File: goon_test.go (hunk 3 above)
Comment (yuichi1004) [reply goon_test.go]: Fixed dd18dfdcf507a3c9df03d209d940e71b82833e10...1c5058d623f2f06c29ab918d5b2e79bf6c916f53

**Comment by yuichi1004** [comment]
@xStrom I fixed where you made comments. Please take a look.

While I was working on the test fix, I found a bug on caching slice. That's why I ended up with like 
1c5058d.

I created an issue since this is not related to PLS and we would like to discuss and to fix in another PR I guess.

Issue:
https://github.com/mjibson/goon/issues/76

**Comment by xStrom** [comment]
Reusing `dst` isn't a good idea, but not because of a bug as I explained in #76. Using a factory function is a decent solution but I think having 8 different entries in `testcase` is even better, because then we keep all the test case configuration together in once place (in the struct).

Currently you did:

```go
for _, tt := range testcase {
    for _, IgnoreFieldMismatch = range []bool{true, false} {
```

The problem is that this `IgnoreFieldMismatch` is a local variable now inside the for loop scope. It doesn't actually modify the global `IgnoreFieldMismatch` variable. I know, I know - global variables can be tricky and confusing, and this is a good example of mistakes that happen with them.

A good solution would be for example:

```go
for _, ifm = range []bool{true, false} {
    testcase = append(testcase, 
        {
            name:                fmt.Sprintf("NormalCache -> NormalCache (IgnoreFieldMismatch:%v)", ifm),
            ignoreFieldMismatch: ifm,
            src:                 &MigrationA{Parent: parentKey, Id: 1},
            dst:                 &MigrationB{Parent: parentKey, Identification: 1},
        })
    // .. and the 3 other appends as well, with ignoreFieldMismatch based on loop variable ifm
}
```

This way we would keep the test configuration together and we would also have a fresh `dst` for every case, so we wouldn't run into the slice reuse issue.

**Comment by yuichi1004** [comment]
@xStrom I'm sorry to took this long time. I applied your feedback.
https://github.com/mjibson/goon/pull/69/commits/972a214dfb23a80f33a7decc16dd236e4f911eb9

File: goon_test.go [hunk 8]
```diff
@@ -892,92 +948,142 @@ func TestMigration(t *testing.T) {
...
-
-		// Clear all the caches
-		g.FlushLocalCache()
-		memcache.Flush(c)
-
-		// Get it back, so it's in the cache
-		migA = &MigrationA{Parent: parentKey, Id: 1}
-		if err := g.Get(migA); err != nil {
-			t.Errorf("Unexpected error on Get: %v", err)
+	for _, IgnoreFieldMismatch := range []bool{true, false} {
```
Comment (xStrom) [inline goon_test.go]: This loop declares a local scope variable called `IgnoreFieldMismatch` which hides the global one. I think the easiest fix is to change `:=` to `=`.

**Comment by xStrom** [comment]
After we get this one final minor issue fixed (`:=` to `=`), I think this PR is in a good enough state to get merged. Unless you still want to add something more here.

**Comment by xStrom** [comment]
Actually one other small thing. Could you also modify the `MemcacheKey` function in goon.go from version g2 to g3. Just to be safe.

```go
// MemcacheKey returns key string of Memcache.
var MemcacheKey = func(k *datastore.Key) string {
	// Versioning, so that incompatible changes to the cache system won't cause problems
	return "g3:" + k.Encode()
}
```

File: goon_test.go (hunk 8 above)
Comment (yuichi1004) [reply goon_test.go]: Well, that's a silly one...
Let me fix that quickly.

Comment (yuichi1004) [reply goon_test.go]: Fixed by 7bb8fe0

**Comment by yuichi1004** [comment]
@xStrom I fixed what you pointed out.

**Comment by xStrom** [comment]
Great! I will merge this tomorrow (1 year anniversary of the PR 😄) unless someone objects.

**Pull request merged**